scripts/valhalla_build_extract

```diff
@@ -0,0 +1,158 @@
+#!/usr/bin/env python3
+
+import argparse
+from io import BytesIO
+import json
+import logging
+from pathlib import Path
+import struct
+import sys
+import tarfile
+from tarfile import BLOCKSIZE
+from typing import Dict
+
+description = """Builds a tar extract from the tiles in mjolnir.tile_dir to the path specified in mjolnir.tile_extract."""
+
+# "=" prefix means native byte order with standard size and no alignment:
+# https://docs.python.org/3/library/struct.html#byte-order-size-and-alignment
+STRUCT_FORMAT = '=LQ'
+STRUCT_SIZE = struct.calcsize(STRUCT_FORMAT)
+INDEX_FILE = "index.bin"
+TAR_ARCNAME_DIR = Path('valhalla_tiles')
+
+parser = argparse.ArgumentParser(description=description)
+parser.add_argument("-c", "--config", help="Absolute or relative path to the Valhalla config JSON.", type=Path)
+parser.add_argument("-v", "--verbosity", help="Accumulative verbosity flags; -v: INFO, -vv: DEBUG", action='count', default=0)
+
+# set up the logger basics
+LOGGER = logging.getLogger(__name__)
+handler = logging.StreamHandler()
+handler.setFormatter(logging.Formatter("%(asctime)s %(levelname)5s: %(message)s"))
+LOGGER.addHandler(handler)
+
+
+def get_padded_tar_size(data_size: int) -> int:
+    """Returns the tar-internal file size, incl. header and NULL padding"""
+    #      <header>    <-data->    <-------------------NULL padding---------------->
+    return BLOCKSIZE + data_size + (BLOCKSIZE - (data_size % BLOCKSIZE or BLOCKSIZE))
+
+
+def get_tile_info(in_path: Path, tiles: dict):
+    """Recursively fills the passed map with tile path & size info"""
+    for p in in_path.iterdir():
+        if p.is_dir():
+            get_tile_info(p, tiles)
+        if p.is_file():
+            tiles[p.resolve()] = get_padded_tar_size(p.stat().st_size)
+
+
+def create_extract(tiles_fp: Path, extract_fp: Path):
+    if tiles_fp.exists() and tiles_fp.is_dir():
+        # collect the tile paths and file sizes
+        files_stats: Dict[Path, int] = dict()
+        get_tile_info(tiles_fp, files_stats)
+        tiles_count = abs(len(list(filter(lambda f: not f.name.endswith('.gph'), files_stats))) - len(files_stats))
+
+        index_size = STRUCT_SIZE * tiles_count
+        index_bin = b''
+        for tile_path, tile_size in files_stats.items():
+            # extract level and ID portion to create a GraphId
+            # tile_path is the absolute path to the tile
+            try:
+                tile_path = str(tile_path.relative_to(tiles_fp))[:-4]
+                level, idx = tile_path.split('/', 1)
+                tile_id = int(level) | (int(idx.replace('/', '')) << 3)
+            except ValueError:
+                # ignore invalid tile ids, so other files can be tarred
+                continue
+
+            LOGGER.debug(f"Tile {tile_path} (ID {tile_id}) with padded tar size: {tile_size}")
+
+            # add 4 & 8 byte ints to the index file
+            index_bin += struct.pack(STRUCT_FORMAT, tile_id, tile_size)
```

Comment:
instead of byte offset, we store the padded file size, which I think is easier for the graphreader logic